### PR TITLE
feat(repo): add balance-monitor

### PR DESCRIPTION
**Pull request by xiaodino**
Initial PR for ETH / Token balance monitoring service

### CODE REVIEW

File: packages/balance-monitor/cmd/utils/subcommand.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+		// _, startMetrics := metrics.Serve(ctx, c)
+
+		// go func() {
+		// 	if err := startMetrics(); err != nil {
+		// 		slog.Error("Starting metrics server error", "error", err)
+		// 	}
+		// }()
+
+		slog.Info("Starting Taiko relayer application", "name", app.Name())
```
Comment (davidtaikocha) [inline packages/balance-monitor/cmd/utils/subcommand.go:38]:
```suggestion
		slog.Info("Starting Taiko balance monitor application", "name", app.Name())
```

File: packages/balance-monitor/cmd/main.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+func main() {
+	app := cli.NewApp()
+
+	log.SetOutput(os.Stdout)
+	// attempt to load a .env file to overwrite CLI flags, but allow it to not
+	// exist.
+
+	app.Name = "Taiko Relayer"
```
Comment (davidtaikocha) [inline packages/balance-monitor/cmd/main.go:22]:
```suggestion
	app.Name = "Taiko Balance Monitor"
```

File: packages/balance-monitor/balance.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
+package main
```
Comment (cyberhorsey) [inline packages/balance-monitor/balance.go:1]: what is this file?

File: packages/balance-monitor/cmd/main.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	app := cli.NewApp()
+
+	log.SetOutput(os.Stdout)
+	// attempt to load a .env file to overwrite CLI flags, but allow it to not
+	// exist.
+
+	app.Name = "Taiko Balance Monitor"
+	app.Usage = "The taiko relayer software command line interface"
+	app.Copyright = "Copyright 2021-2024 Taiko Labs"
+	app.Description = "Bridge relayer implementation in Golang for Taiko protocol"
```
Comment (cyberhorsey) [inline packages/balance-monitor/cmd/main.go:25]: not bridge relayer, but balance monitor (also in Usage above)

File: packages/balance-monitor/cmd/main.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	app.Authors = []*cli.Author{{Name: "Taiko Labs", Email: "info@taiko.xyz"}}
+	app.EnableBashCompletion = true
+
+	// All supported sub commands.
+	app.Commands = []*cli.Command{
+		{
+			Name:        "balance-monitor",
+			Flags:       flags.CommonFlags,
+			Usage:       "Starts the balance monitor oftware",
+			Description: "Taiko relayer balance monitro",
```
Comment (cyberhorsey) [inline packages/balance-monitor/cmd/main.go:35]: `monitro` typo

File: packages/relayer/docker-compose/docker-compose.yml [hunk 6]
```diff
@@ -7,7 +7,7 @@ services:
     restart: always
     environment:
       - MYSQL_DATABASE=relayer
-      - MYSQL_ROOT_PASSWORD=<password>
+      - MYSQL_ROOT_PASSWORD=passw00d
```
Comment (cyberhorsey) [inline packages/relayer/docker-compose/docker-compose.yml:10]: not supposed to use even default passwords according to Daniel

File: packages/balance-monitor/balance-monitor/balance_monitor.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+	SuggestGasPrice(ctx context.Context) (*big.Int, error)
+	SuggestGasTipCap(ctx context.Context) (*big.Int, error)
+}
+
+type BalanceMonitor struct {
+	l1EthClient        ethClient
+	l2EthClient        ethClient
+	Addresses          []common.Address
+	ERC20Addresses     []common.Address
+	Interval           int
```
Comment (cyberhorsey) [inline packages/balance-monitor/balance-monitor/balance_monitor.go:48]: interval, addresses, ERC20Addresses should all be locally scoped (lower case)

File: packages/balance-monitor/balance-monitor/balance_monitor.go [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/urfave/cli/v2"
+	"golang.org/x/exp/slog"
+)
+
+type ethClient interface {
+	BalanceAt(ctx context.Context, account common.Address, blockNumber *big.Int) (*big.Int, error)
```
Comment (cyberhorsey) [inline packages/balance-monitor/balance-monitor/balance_monitor.go:24]: The vast majority of these methods are unused, and dont need to be stubbed for the interface here. we only need BalanceAt and maybe some others that you actually use.

File: packages/balance-monitor/balance-monitor/balance_monitor.go [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+	slog.Info(fmt.Sprintf("%s ERC-20 Balance", clientLabel), "tokenAddress", tokenAddress.Hex(), "address", holderAddress.Hex(), "balance", tokenBalanceFloat)
+}
+
+const erc20ABI = `[{"constant":true,"inputs":[{"name":"_owner","type":"address"}],"name":"balanceOf","outputs":[{"name":"balance","type":"uint256"}],"type":"function"},{"constant":true,"inputs":[],"name":"decimals","outputs":[{"name":"","type":"uint8"}],"type":"function"}]`
+
+type ERC20 interface {
+	BalanceOf(opts *bind.CallOpts, account common.Address) (*big.Int, error)
+}
+
+func (b *BalanceMonitor) GetEthBalance(ctx context.Context, client ethClient, address common.Address) (*big.Int, error) {
```
Comment (cyberhorsey) [inline packages/balance-monitor/balance-monitor/balance_monitor.go:158]: this method should be private

File: packages/balance-monitor/balance-monitor/balance_monitor.go [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+	balance, err := client.BalanceAt(ctx, address, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return balance, nil
+}
+
+func (b *BalanceMonitor) GetERC20Balance(ctx context.Context, client ethClient, tokenAddress, holderAddress common.Address) (*big.Int, error) {
```
Comment (cyberhorsey) [inline packages/balance-monitor/balance-monitor/balance_monitor.go:167]: this method should be private

File: packages/balance-monitor/balance-monitor/balance_monitor.go [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+
+	balance, ok := result[0].(*big.Int)
+	if !ok {
+		return nil, fmt.Errorf("unexpected type for balanceOf result")
+	}
+
+	return balance, nil
+}
+
+func (b *BalanceMonitor) GetERC20Decimals(ctx context.Context, client ethClient, tokenAddress common.Address) (uint8, error) {
```
Comment (cyberhorsey) [inline packages/balance-monitor/balance-monitor/balance_monitor.go:196]: this method shoulod be private

**Review by cyberhorsey — CHANGES_REQUESTED** [review]

File: packages/relayer/docker-compose/docker-compose.yml [hunk 12]
```diff
@@ -29,7 +29,7 @@ services:
     container_name: prometheus
     restart: always
     volumes:
-      - ./prometheus.yml:/etc/prometheus/prometheus.yml
+      - ./prometheus/prometheus.yml:/etc/prometheus/prometheus.yml
```
Comment (cyberhorsey) [inline packages/relayer/docker-compose/docker-compose.yml:32]: is this change intended for the relayer, and not a new folder for balance-monitor?

**Review by cyberhorsey — CHANGES_REQUESTED** [review]

File: packages/balance-monitor/balance-monitor/balance_monitor.go (hunk 8 above)
Comment (xiaodino) [reply packages/balance-monitor/balance-monitor/balance_monitor.go:24]: Removed unused methods. The rest is needed (I just check them one by one by removing it)

File: packages/balance-monitor/balance-monitor/balance_monitor.go (hunk 9 above)
Comment (xiaodino) [reply packages/balance-monitor/balance-monitor/balance_monitor.go]: Updated.

File: packages/balance-monitor/balance-monitor/balance_monitor.go (hunk 7 above)
Comment (xiaodino) [reply packages/balance-monitor/balance-monitor/balance_monitor.go]: Updated.

File: packages/relayer/docker-compose/docker-compose.yml (hunk 6 above)
Comment (xiaodino) [reply packages/relayer/docker-compose/docker-compose.yml]: Reverted.

File: packages/balance-monitor/cmd/main.go (hunk 5 above)
Comment (xiaodino) [reply packages/balance-monitor/cmd/main.go]: Updated.

File: packages/balance-monitor/cmd/main.go (hunk 4 above)
Comment (xiaodino) [reply packages/balance-monitor/cmd/main.go]: Updated.

File: packages/balance-monitor/balance.go (hunk 3 above)
Comment (xiaodino) [reply packages/balance-monitor/balance.go]: Removed. It was a test file.

File: packages/balance-monitor/cmd/main.go (hunk 2 above)
Comment (xiaodino) [reply packages/balance-monitor/cmd/main.go]: Updated.

File: packages/balance-monitor/cmd/utils/subcommand.go (hunk 1 above)
Comment (xiaodino) [reply packages/balance-monitor/cmd/utils/subcommand.go]: Updated.

File: packages/balance-monitor/balance-monitor/balance_monitor.go (hunk 10 above)
Comment (xiaodino) [reply packages/balance-monitor/balance-monitor/balance_monitor.go]: Updated.

File: packages/balance-monitor/balance-monitor/balance_monitor.go (hunk 11 above)
Comment (xiaodino) [reply packages/balance-monitor/balance-monitor/balance_monitor.go]: Updated.

File: packages/relayer/docker-compose/docker-compose.yml (hunk 12 above)
Comment (xiaodino) [reply packages/relayer/docker-compose/docker-compose.yml]: Updated.

**Review by cyberhorsey — APPROVED** [review]

**Comment by cyberhorsey** [comment]
@xiaodino i think we need to add a workflow to build this image in the .github folder.

**Comment by xiaodino** [comment]
> @xiaodino i think we need to add a workflow to build this image in the .github folder.

I will add it in a follow up PR.

**Review by dantaik — APPROVED** [review]
